### PR TITLE
Use linenoise instead of liner for interactive "cockroach sql".

### LINKED ISSUE #4134: multi-line editing in cockroach --sql is messed up.
If you type a command on 2 lines, hit enter, then hit up-arrow to get back the last command, all sorts of things are broken. For example, you can only edit the 2nd line of the command; pressing any key (including up arrow again) prints a completely new prompt line for no reason.

Also hitting enter on a line with no other characters initiates a multi-line command. What it should do is just give you a fresh prompt line.

### PR DESCRIPTION
**Pull request by knz**
Fixes #4134.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4209)

### CODE REVIEW

**Comment by knz** [comment]
Note that the circleci checks fail because the circleci conf does not include readline headers and libs. This needs to be addressed as well, however I could not figure out how myself.

**Comment by tamird** [comment]
Yikes, this doesn't look like a step in the right direction to me. CI failed to find the readline header - this is generally indicative of the shenanigans we'll need to go through to get this working.

> persistent history

liner supports this, we're just not using it.

> history navigation and editing

liner supports searching history, though not editing.

> support for user-global ~/.inputrc
keyboard macros
and more! check the man page readline(3)

Maybe this is nice, but I don't think it's worth the added dependency.

**Comment by knz** [comment]
@tamird don't let the circleci issue come in the way of evaluating this PR on its own merits. This would be a tail wagging its dog.

Readline is an industry standard and is rock-solid. Using it directly fixes #4134. In comparison liner is relatively recent and untested, has less support for terminals and fixing #4134 will require teaching liner about more terminal semantics than its architecture is ready to deal with atm (= a major rehaul of the library).

**Comment by tamird** [comment]
> @tamird don't let the circleci issue come in the way of evaluating this PR on its own merits. This would be a tail wagging its dog.

This view is overly simplistic. I'm not keen on introducing a `go get`-incompatible dependency. This complicates the developer story for all contributors to the project.

> than its architecture is ready to deal with atm (= a major rehaul of the library).

Is this true? AFAICT we haven't reported the issue upstream at all.

**Comment by knz** [comment]
I did report the issue upstream https://github.com/peterh/liner/issues/68.

**Comment by knz** [comment]
Also I'd rather have an external dependency to a well-known, pervasive library (you likely won't find libreadline less available than GNU make, which we already require) than sustain the general case of NIH syndrome found throughout Go world. This is not Java!

**Comment by tamird** [comment]
Actually, in this regard, it pretty much is :)

**Comment by bdarnell** [comment]
Separate from the technical considerations, `readline` is licensed under the GPL v3, which means we can't use it in our apache-licensed product. `editline` is a BSD-licensed alternative, although a naive use of `editline` would have the same build-related headaches. I'm not willing to complicate the build process for better editing in the SQL shell; we need either a pure-go solution or a cgo-based build wrapper as we've done with rocksdb, protobuf, etc.

**Comment by nvanbenschoten** [comment]
I had previously looked into `liner` alternatives and ran into [chzyer/readline](https://github.com/chzyer/readline). It's MIT-licensed and a pure Go implementation with decent adoption. Might be worth a look as well.

**Comment by tamird** [comment]
Ooh, that's nice.

**Comment by knz** [comment]
I tried that one first but it didn't let me control when to save lines in history so it didn't work well with multi line statements. But perhaps we can submit a simple patch upstream.

Tamir Duberstein <notifications@github.com> schreef op 8 februari 2016 00:21:27 CET:
>Ooh, that's nice.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/cockroachdb/cockroach/pull/4209#issuecomment-181139334

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by knz** [comment]
I went for linenoise instead of chzyer/readline. The latter is a code mess that does not allow to set history separately from reading a line.

**Comment by tamird** [comment]
merge conflict :lock:

**Comment by knz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/sql.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2BJRDFUJCFG89qYc_:-KA2BJRE7w60z-oiJRdy:457323815) ([raw file](https://github.com/cockroachdb/cockroach/blob/7722dbb3318eaf0b3a8d0ac2a6c95beed3e1b190/cli/sql.go#L77)):*
linenoise already does this, that's why the comment is going away.

---

*[cli/sql.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2BNBkpTdQTO0YglAZ:-KA2BNBl_N9JN5eeq_aC:-1815393975) ([raw file](https://github.com/cockroachdb/cockroach/blob/7722dbb3318eaf0b3a8d0ac2a6c95beed3e1b190/cli/sql.go#L79)):*
This comment is an atavism from when there was no sql -e. The SQL tests now use sql -e.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[cli/sql.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2Bk5-Q-0H_b3MecWZ:-KA2Bk5-Q-0H_b3MecW_:-793749798) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L76)):*
remove this comment.

---

*[cli/sql.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2BpPBJuE3HFL_KUnd:-KA2BpPBJuE3HFL_KUne:1809377498) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L78)):*
The right way to get the home directory is using [os/user.Current](https://godoc.org/os/user#Current):
```
import "os/user"

u, err := user.Current
...
u.HomeDir
```

---

*[cli/sql.go, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CB0A8eA1ddQSMUlZ:-KA2CB0A8eA1ddQSMUl_:-793749798) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L82)):*
remove this comment.

---

*[cli/sql.go, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CH-C3Gs0Ypt0mLKx:-KA2CH-C3Gs0Ypt0mLKy:1821824539) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L95)):*
Why not keep the old code? Also, printing "Input terminated!" is needlessly noisy.

---

*[cli/sql.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CUjyvXyaUElE7a9t:-KA2CUjyvXyaUElE7a9u:-959087981) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L124)):*
seems like you can do this just once with a `defer`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[cli/sql.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2Bk5-Q-0H_b3MecWZ:-KA2EegOCKLpfsgNyjsF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L76)):*
Done.

---

*[cli/sql.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2BpPBJuE3HFL_KUnd:-KA2Eg2u_fbrFhJnukdc:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L78)):*
Done.

---

*[cli/sql.go, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CB0A8eA1ddQSMUlZ:-KA2Eh5St4-tmeZGgchJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L82)):*
Done.

---

*[cli/sql.go, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CH-C3Gs0Ypt0mLKx:-KA2EiV2Gy2L7k0_cES7:1326538832) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L95)):*
See the new comment.

---

*[cli/sql.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CUjyvXyaUElE7a9t:-KA2Eksb-irN_9WahCAu:1326538832) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L124)):*
See the new comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

File: cli/sql.go [hunk 1]
```diff
@@ -19,12 +19,13 @@ package cli
...
 	"fmt"
-	"io"
 	"net/url"
 	"os"
+	"os/user"
+	"path"
 	"strings"
 
-	"github.com/peterh/liner"
+	"github.com/GeertJohan/go.linenoise"
```
Comment (petermattis) [inline cli/sql.go]: That capitalization is so eye catching. Not in a good way. :(

**Comment by petermattis** [comment]
LGTM

---

Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-:-KA2L4YAEw0cz9189hbM:676834233)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2LFOibe3KLBNT2BVm:2076263902) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
It's the only decent wrapper I found for linenoise. You think we should fork it under cockroachdb?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2Lg8Y9Ut2c6rrhjBQ:1775699428) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
I think you can also just lowercase `GeertJohan`. You'll have to rename your local copy, though. Or you can leave it be, it's just odd to see.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2MHod_9w_QqLlGx9u:1263827779) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
I tried changing the case, and gofmt cases it back up based on the directory name.
If I rename it in the filesystem, how do we go about "go get" and GLOCK?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2N0PqIh8nYFzInCSf:-1397149203) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
`go get` should work on the lowercase name. And once it is lowercase in the filesystem, GLOCK should save the right name. I think. But don't spend a lot of time on this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2NXa_N7BN2BNRND6P:2089070394) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
Ok right that worked. Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by bdarnell** [comment]
Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2NQiYnMENmOs9JrSS:-1997944599) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
I think we should leave this with the upstream capitalization (we already have github.com/VividCortex/ewma). Someone may have already checked this package out in their GOPATH under the mixed-case name and then things can get weird (remember that at least the OSX is case-insensitive but case-preserving; lots of tools don't handle renames that only differ in case gracefully). 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 28 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-K9x_9fUE5wFOJyYF5Wi-r2-28:-KA2Oj-nFNP9k6ZDAO-h:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9324cac5a848504f13ecbafaa3a3678b057d03b/cli/sql.go#L28)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[cli/sql.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CUjyvXyaUElE7a9t:-KA2RmyKSrspZEs0f8Qc:-1038856393) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L124)):*
`defer`s are run even on `panic`.

---

*[cli/sql.go, line 58 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2S7d6dlbFgIl66QUl:-KA2S7d6dlbFgIl66QUm:1758399671) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L58)):*
is this comment useful?

---

*[cli/sql.go, line 79 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SKxaDSjz2pxFl2Lu:-KA2SKxaDSjz2pxFl2Lv:-1419550079) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L79)):*
we shouldn't ignore this error. log it at a high verbosity

---

*[cli/sql.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SA6yDr3_dSQM0h7O:-KA2SA6yDr3_dSQM0h7P:1862925852) ([raw file](https://github.com/cockroachdb/cockroach/blob/7722dbb3318eaf0b3a8d0ac2a6c95beed3e1b190/cli/sql.go#L81)):*
moving these `var`s is just line noise, can you leave them where they are?

---

*[cli/sql.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SFomfQS9xxxqqVMw:-KA2SFomfQS9xxxqqVMx:-1419550079) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L81)):*
we shouldn't ignore this error. log it at a high verbosity

---

*[cli/sql.go, line 93 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2RfhalIw9YrodzcS9:-KA2RfhbYOzRY2ouL7Vy:-1937250472) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L93)):*
huh? a string?

---

*[cli/sql.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SRipIVXiSzMcTJgP:-KA2SRipIVXiSzMcTJgQ:-1419550079) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L128)):*
we shouldn't ignore this error. log it at a high verbosity

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[cli/sql.go, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2CUjyvXyaUElE7a9t:-KA4ZIm2_6MIw4Yx4QCv:-1925644738) ([raw file](https://github.com/cockroachdb/cockroach/blob/460528ecc2a7b96e4aef75002fcf3168b481042a/cli/sql.go#L124)):*
TIL, thanks. However see the new comment.

---

*[cli/sql.go, line 58 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2S7d6dlbFgIl66QUl:-KA4WoLZUNhdSotU0Sgf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L58)):*
Done.

---

*[cli/sql.go, line 79 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SKxaDSjz2pxFl2Lu:-KA4Y3fMTEH58W-a5rF-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L79)):*
Done.

---

*[cli/sql.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SFomfQS9xxxqqVMw:-KA4YjkeMPHPh9ItpIYO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L81)):*
Done.

---

*[cli/sql.go, line 81 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SA6yDr3_dSQM0h7O:-KA4YzdsVovk0sEEwAh5:-1343796116) ([raw file](https://github.com/cockroachdb/cockroach/blob/7722dbb3318eaf0b3a8d0ac2a6c95beed3e1b190/cli/sql.go#L81)):*
Some but not all. Done.

---

*[cli/sql.go, line 93 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2RfhalIw9YrodzcS9:-KA4ZH9g2ONkOJC8GnC6:-888365374) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L93)):*
Clarified.

---

*[cli/sql.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA2SRipIVXiSzMcTJgP:-KA4aqZQfywNJ7XqZEnW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb88b9bc572f5c5b3a6f2c59c2f95aa7b8e71e37/cli/sql.go#L128)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 57 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4tmTINBaJEi-_OT8b:-KA4tmTJzxDRkO8pz71n:1283409683) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L57)):*
there's no reason to move this, can you put it back?

---

*[cli/sql.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4tqi7q087hyjORciw:-KA4tqi7q087hyjORcix:449548612) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L86)):*
The methods Info/Warning/Error etc do not correspond to verbosity, only severity. To log at high verbosity you should do something like
```go
if log.V(2) {
....
}
```

Also, in this case you need the `Warningf` variant.

---

*[cli/sql.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4uBZKfKQj54UfwFDf:-KA4uBZL3SpXIRGZyt-S:-1682615444) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L89)):*
you need to use `histFile =` - you're currently shadow the out scope, not overwriting it.

---

*[cli/sql.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4uJ98xYCwVvx79NHg:-KA4uJ98xYCwVvx79NHh:-414689656) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L90)):*
you can use `if err := ...; err != nil {` here (which will let you move the `var err` down to where it used to be).

---

*[cli/sql.go, line 109 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4ua-VC5vhnDgNENui:-KA4ua-WyDaKcJ6A-OY-:-187080173) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L109)):*
The error you mention here is an exported variable (https://github.com/GeertJohan/go.linenoise/blob/1918ff89d61336553294cc577e9a32b647732fff/linenoise.go#L17)

Instead of this comment, can you:
```
if err != linenoise. KillSignalError {
  exitErr = err
}
``
?

---

*[cli/sql.go, line 151 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4v2og8VLBX5XKO42p:-KA4v2og8VLBX5XKO42q:-890587922) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L151)):*
ditto `if err := ...; err != nil {`

---

*[cli/sql.go, line 153 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4v5VSGHS-F1ysNGu4:-KA4v5VSGHS-F1ysNGu5:-1089398266) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L153)):*
ditto re verbosity and `Warningf`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*
<!-- Sent from Reviewable.io -->

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/sql.go, line 57 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4tmTINBaJEi-_OT8b:-KA50h1Xsd92jpEXifpM:-1464448035) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L57)):*
Or even better remove it :stuck_out_tongue: 

---

*[cli/sql.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4tqi7q087hyjORciw:-KA4yOnAQ_m4Io1A80Ny:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L86)):*
Done.

---

*[cli/sql.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4uBZKfKQj54UfwFDf:-KA4yR_Dz0wFvufGMnCy:862796271) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L89)):*
Oops. Well spotted. Thanks to code review!

---

*[cli/sql.go, line 90 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4uJ98xYCwVvx79NHg:-KA4yeXB3HTQqM_bSToR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L90)):*
Done.

---

*[cli/sql.go, line 109 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4ua-VC5vhnDgNENui:-KA4z7VS-DHT2qMsfKsW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L109)):*
Done.

---

*[cli/sql.go, line 151 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4v2og8VLBX5XKO42p:-KA4zp8OpFQbvRrJpXBg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L151)):*
Done.

---

*[cli/sql.go, line 153 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA4v5VSGHS-F1ysNGu4:-KA4zMbwvfZrax2VGxB-:-1141514743) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5dad0d06511db3fde484fbbe45bc6c3c05c6e2/cli/sql.go#L153)):*
Agreed for Warningf, however for verbosity level I disagree; I think the user will very mch want to know why the sql client cannot write a file in their own home directory.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by tamird** [comment]
LGTM but I think there are some logging issues that new users will find alarming. I encourage you to exercise all the logging paths you've added here at zero-verbosity to make sure it's all sane.

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/sql.go, line 86 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5SU2PRx1m1go3It0l:-KA5SU2QmjHlTrRLlSrf:680169268) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L86)):*
I think this should be log.Info like you have way down below `log.Info("command-line history will not be saved further in this session.")` with the phrasing adjusted to remove "further" and the trailing period

---

*[cli/sql.go, line 91 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5RhMjtHuSpF8cY9id:-KA5RhMjtHuSpF8cY9ie:-301394329) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L91)):*
this should definitely be logged at V(1) or higher; currently every new user will see this (right?).

---

*[cli/sql.go, line 101 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5RDDyidFtXPa9MUxM:-KA5RDDyidFtXPa9MUxN:316535683) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L101)):*
I like expressions too, but is this really more readable? At the very least, allocate this map just once.

---

*[cli/sql.go, line 148 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5Sd91HufNoGmH86st:-KA5Sd921RZHn9utrx-r:690744283) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L148)):*
no periods in log messages

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-:-KA5SwRwOr_UfrWTmmS1:-1136846821)*

**Comment by knz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[cli/sql.go, line 86 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5SU2PRx1m1go3It0l:-KA5dSsWDbq1Yh8Z2kzd:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L86)):*
Done.

---

*[cli/sql.go, line 91 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5RhMjtHuSpF8cY9id:-KA5dUVCXeFyhtrgx_uR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L91)):*
Done.

---

*[cli/sql.go, line 101 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5RDDyidFtXPa9MUxM:-KA5hTD_GqsmglcyXjZX:-820920943) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L101)):*
Simplified.

---

*[cli/sql.go, line 148 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-KA5Sd91HufNoGmH86st:-KA5hVJCEo3vFbeKyFE_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/62dd528f2da7971e6798e5521c4be10a81899591/cli/sql.go#L148)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209)*

**Comment by knz** [comment]
You were right, I fixed them by manually testing.

---

Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4209#-:-KA5i1tQkLz-E2yNPQkw:1401174188)*

**Comment by chzyer** [comment]
@knz I added a option to disable auto save history and enable to save history manually.
Here is the [demo code](https://github.com/chzyer/readline/blob/master/example/readline-multiline/readline-multiline.go). Is there any change to use [readline](https://github.com/chzyer/readline)？ (・ω・)ノ

**Comment by knz** [comment]
Hey @chzyer o/ 
I will have a look, of course feel free to suggest a PR too :grin: I'd definitely champion that.

**Pull request merged**